File: app/pipeline/supa.py
"""Supabase helpers for the flat-2D studio pipeline (requests-based, secret key).

Tables: assets, videos, video_beats, jobs, brands.  Buckets: assets, videos (both public).
Job queue: insert_job() to enqueue, claim_job() (RPC) to atomically claim, update_job() to finish.
"""
import mimetypes, uuid
from pathlib import Path
import requests
from dotenv import dotenv_values
import os
_env = {}
try: _env = dotenv_values('/Users/rudra/OpenMontage/.env')
except Exception: pass
URL = (os.environ.get('SUPABASE_URL') or _env.get('SUPABASE_URL')).rstrip('/')
KEY = os.environ.get('SUPABASE_SECRET') or _env.get('SUPABASE_SECRET')
HDRS = {'apikey': KEY, 'Authorization': f'Bearer {KEY}'}


def _rest(method, path, params=None, body=None, prefer=None):
    h = dict(HDRS, **{'Content-Type': 'application/json'})
    if prefer: h['Prefer'] = prefer
    r = requests.request(method, f'{URL}/rest/v1/{path}', headers=h, params=params, json=body, timeout=60)
    r.raise_for_status()
    return r.json() if r.text.strip() else None


# ---- generic row helpers ----
def insert(table, row):
    """Insert one row, return the created row (with id)."""
    return _rest('POST', table, body=row, prefer='return=representation')[0]

def update(table, row_id, patch):
    """PATCH one row by id, return the updated row (or None if id not found)."""
    d = _rest('PATCH', table, params={'id': f'eq.{row_id}'}, body=patch, prefer='return=representation')
    return d[0] if d else None

def select(table, **filters):
    """Simple select, e.g. select('jobs', status='eq.queued'). Returns list of rows."""
    return _rest('GET', table, params=filters) or []

# ---- named helpers (videos / video_beats / jobs) ----
def insert_video(row):        return insert('videos', row)
def update_video(vid, patch): return update('videos', vid, patch)
def insert_beat(row):         return insert('video_beats', row)
def update_beat(bid, patch):  return update('video_beats', bid, patch)
def insert_job(row):          return insert('jobs', row)
def update_job(jid, patch):   return update('jobs', jid, patch)


def claim_job():
    """Atomically claim the next queued job via the claim_job() RPC.
    Returns the claimed job row (dict) or None if the queue is empty."""
    d = _rest('POST', 'rpc/claim_job', body={})
    if isinstance(d, list):
        return d[0] if d else None
    return d or None


# ---- storage ----
def _upload(bucket, path, filepath, mime):
    h = dict(HDRS, **{'Content-Type': mime, 'x-upsert': 'true'})
    with open(filepath, 'rb') as f:
        r = requests.post(f'{URL}/storage/v1/object/{bucket}/{path}', headers=h, data=f, timeout=600)
    r.raise_for_status()
    return path

def public_url(path, bucket='assets'):
    return f'{URL}/storage/v1/object/public/{bucket}/{path}'

def upload_asset(filepath, kind, title, tags=[], style=None, brand_id=None, duration_s=None, meta={}, cost=0):
    """Upload a file to the 'assets' bucket at <kind>/<uuid>.<ext> and insert an assets row.
    Returns the inserted assets row.
    DB check constraint — kind must be one of: still, clip, vo, music, sfx, logo,
    intro, outro, font, char_ref, graphic, video."""
    filepath = Path(filepath)
    ext = filepath.suffix.lstrip('.').lower() or 'bin'
    path = f'{kind}/{uuid.uuid4()}.{ext}'
    mime = mimetypes.guess_type(filepath.name)[0] or 'application/octet-stream'
    _upload('assets', path, filepath, mime)
    return insert('assets', {'kind': kind, 'title': title, 'tags': tags, 'style': style,
                             'brand_id': brand_id, 'storage_path': path, 'mime': mime,
                             'duration_s': duration_s, 'meta': meta, 'cost_usd': cost})

def upload_video(filepath, video_id):
    """Upload a finished render to the 'videos' bucket at <video_id>.<ext>.
    Returns the storage path (use public_url(path, 'videos') for the URL)."""
    filepath = Path(filepath)
    ext = filepath.suffix.lstrip('.').lower() or 'mp4'
    mime = mimetypes.guess_type(filepath.name)[0] or 'video/mp4'
    return _upload('videos', f'{video_id}.{ext}', filepath, mime)

File: recorder/supa.mjs
// Minimal Supabase helper for the recorder service (service key, REST).
// Browsercast jobs are 'videos' rows the Python worker never sees (no 'jobs'
// row), so the two services never fight over the queue.
import { readFileSync } from 'fs';
import { randomUUID } from 'crypto';

const BASE = (process.env.SUPABASE_URL || '').replace(/\/$/, '');
const KEY = process.env.SUPABASE_SECRET;
if (!BASE || !KEY) throw new Error('SUPABASE_URL / SUPABASE_SECRET not set');
const H = { apikey: KEY, Authorization: `Bearer ${KEY}` };

async function rest(method, path, { params, body, prefer } = {}) {
  const u = new URL(`${BASE}/rest/v1/${path}`);
  if (params) for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  const h = { ...H, 'Content-Type': 'application/json' };
  if (prefer) h.Prefer = prefer;
  const r = await fetch(u, { method, headers: h, body: body ? JSON.stringify(body) : undefined });
  if (!r.ok) throw new Error(`supabase ${method} ${path} ${r.status} ${(await r.text()).slice(0, 200)}`);
  const t = await r.text();
  return t.trim() ? JSON.parse(t) : null;
}

// claim the next queued browsercast video (optimistic lock: only one worker wins)
export async function claimBrowsercast() {
  const rows = await rest('GET', 'videos', {
    params: { status: 'eq.queued', 'progress->>via': 'in.(browsercast,upload)',
      order: 'created_at.asc', limit: '1', select: 'id,title,brand_id,progress' },
  });
  if (!rows || !rows.length) return null;
  const v = rows[0];
  const claimed = await rest('PATCH', 'videos', {
    params: { id: `eq.${v.id}`, status: 'eq.queued' },
    body: { status: 'running' }, prefer: 'return=representation',
  });
  return claimed && claimed.length ? v : null;   // empty => another worker took it
}

export async function updateVideo(id, patch) {
  return rest('PATCH', 'videos', { params: { id: `eq.${id}` }, body: patch, prefer: 'return=representation' });
}

// on startup, requeue browsercast videos left 'running' by a crashed/restarted
// instance (only one recorder runs at a time, so any 'running' one is an orphan)
export async function recoverOrphans() {
  const rows = await rest('PATCH', 'videos', {
    params: { status: 'eq.running', 'progress->>via': 'in.(browsercast,upload)' },
    body: { status: 'queued' }, prefer: 'return=representation' });
  return rows ? rows.length : 0;
}

export async function getBrand(id) {
  if (!id) return null;
  const rows = await rest('GET', 'brands', { params: { id: `eq.${id}`, limit: '1' } });
  return rows && rows.length ? rows[0] : null;
}

// download an asset (by id) to a local path via its public URL
export async function downloadAsset(assetId, dest) {
  if (!assetId) return null;
  const rows = await rest('GET', 'assets', { params: { id: `eq.${assetId}`, select: 'storage_path', limit: '1' } });
  if (!rows || !rows.length) return null;
  const r = await fetch(publicUrl(rows[0].storage_path));
  if (!r.ok) throw new Error(`download asset ${assetId} ${r.status}`);
  const { writeFileSync } = await import('fs');
  writeFileSync(dest, Buffer.from(await r.arrayBuffer()));
  return dest;
}

const MIME = { mp4: 'video/mp4', webm: 'video/webm', wav: 'audio/wav', mp3: 'audio/mpeg', png: 'image/png', jpg: 'image/jpeg' };

// upload a file to the public 'assets' bucket and insert an assets row
export async function uploadAsset(filepath, kind, { title, tags = [], brand_id = null, duration_s = null } = {}) {
  const ext = (filepath.split('.').pop() || 'bin').toLowerCase();
  const path = `${kind}/${randomUUID()}.${ext}`;
  const mime = MIME[ext] || 'application/octet-stream';
  const r = await fetch(`${BASE}/storage/v1/object/assets/${path}`, {
    method: 'POST', headers: { ...H, 'Content-Type': mime, 'x-upsert': 'true' },
    body: readFileSync(filepath),
  });
  if (!r.ok) throw new Error(`upload ${r.status} ${(await r.text()).slice(0, 200)}`);
  const row = await rest('POST', 'assets', {
    body: { kind, title, tags, brand_id, storage_path: path, mime, duration_s },
    prefer: 'return=representation',
  });
  return row[0];
}

export function publicUrl(path, bucket = 'assets') {
  return `${BASE}/storage/v1/object/public/${bucket}/${path}`;
}
